### ISSUE TITLE
passing globals only once

### ISSUE THREAD

**Issue by gdkrmr** [HenrikBengtsson/future.apply#11]
I am trying to use `future_lapply` as a drop in replacemant for `parLapplyLB`. I have a `very_large_object` which I pass around once and then work on it with a simple index, `very_large_object` is not changed in the process:

```R
cl <- makeCluster(rep("node1", 64))
clusterExport(cl, "very_large_object")
f <- function(i) do_something_useful(i, very_large_object)
result <- parLapplyLB(cl, 1:1000, f)
```
From the documentation I am under the impression that
```R
plan(tweak(cluster, workers = cl))
result <- future_lapply(1:1000, f)
```
will send my `very_large_object` 1000 times to my workers. Setting `future.globals = FALSE` and sending the object just as above does not work.
Is there a way to avoid this?

**Comment by HenrikBengtsson** [comment]
Hi, when do you do:
```r
plan(cluster, workers = cl)
result <- future_lapply(1:1000, f)
```
the default is that `future_lapply()` will do so called "load balancing".   It will launch exactly `nbrOfWorkers()` (= `length(cl)`) futures and then have each of those futures internally operate on a chunk of `1:1000`.  So, if you have 20 nodes in your cluster `cl`, then `nbrOfWorkers() == 20L` and then each worker will process 1000/20 = 50 elements in `1:1000` (i.e. `1:50`, `51:100`, ..., `951:1000`).  For each of these 20 futures, the `very_large_object` object will be exported.  So, it'll be exported 20 times, not 1000 times.

The above "chunking" of `1:1000` into equally sized sets is controlled by argument `future.scheduling` which defaults to `1`.   If you set it to `future.scheduling = 2`, you'll get twice as many chunks (= futures), i.e. you'll copy `very_large_object` twice the number of times.  If you'd set `future.scheduling = Inf`, then you would get one element per chunk (=future), i.e. you'd copy `very_large_object` 1000 times.

Hope this helps

PS. In R 3.5.0, `parallel::parLapplyLB()` gained argument `chunk.size`.  Similarly to `future.scheduling`, it can be use to control the load balancing ("amount of chunking").  To make things less confusing when going back and forth between `parallel` and `future.apply`, we might introduce a `future.chunk.size` argument as an alternative to `future.scheduling` that works identically to this new `chunk.size` (Issue #7).

**Comment by gdkrmr** [comment]
It seemed like I understood this wrong. Sorry for raising this non-issue. 
I ran into the `parLapplyLB()` but that got fixed in R 3.5.0 and tried different parallel implementations of `lapply` and finally resorted to `clusterApplyLB` --- this was the only case I ever observed where I had to do that, in other cases `clusterApplyLB` is horribly slow. The performance of `future_apply` wasn't really good, it only used 1/4 of the cores effectively for whatever reason, I did not play around with different parameters though.

**Comment by HenrikBengtsson** [comment]
They all rely on basically the same underlying setup of background workers and ways of communicating data back and forth, so it's most likely comes down to how you choose to distribute your load. As a rough approximation, they also all provide similar ways of doing load balancing - but possible with different defaults.  It's unlikely that you'll find one magic solution that will work the best for all scenarios.  Understanding how they work will help when it comes to avoiding inefficient processing.  Regardless, the take-home message is you should be able to get about the same performance regardless of which of those methods you choose.
